hivemq-edge/src/main/java/com/hivemq/api/resources/impl/CombinersResourceImpl.java

```diff
@@ -0,0 +1,179 @@
+package com.hivemq.api.resources.impl;
+
+import com.hivemq.api.errors.AlreadyExistsError;
+import com.hivemq.api.errors.ConfigWritingDisabled;
+import com.hivemq.api.errors.InternalServerError;
+import com.hivemq.api.errors.adapters.AdapterNotFoundError;
+import com.hivemq.combining.CombiningManager;
+import com.hivemq.combining.DataCombiner;
+import com.hivemq.combining.DataCombining;
+import com.hivemq.configuration.info.SystemInformation;
+import com.hivemq.edge.api.CombinersApi;
+import com.hivemq.edge.api.model.Combiner;
+import com.hivemq.edge.api.model.CombinerList;
+import com.hivemq.edge.api.model.DataCombiningList;
+import com.hivemq.persistence.mappings.fieldmapping.Instruction;
+import com.hivemq.util.ErrorResponseUtil;
+import org.apache.commons.lang3.NotImplementedException;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.ws.rs.core.Response;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+
+
+@Singleton
+public class CombinersResourceImpl implements CombinersApi {
+
+
+    private static final Logger log = LoggerFactory.getLogger(CombinersResourceImpl.class);
+
+    private final @NotNull SystemInformation systemInformation;
+    private final @NotNull CombiningManager combiningManager;
+
+    @Inject
+    public CombinersResourceImpl(
+            final @NotNull SystemInformation systemInformation, final @NotNull CombiningManager combiningManager) {
+        this.systemInformation = systemInformation;
+        this.combiningManager = combiningManager;
+    }
+
+    @Override
+    public @NotNull Response addCombiner(final @NotNull Combiner combiner) {
+        if (!systemInformation.isConfigWriteable()) {
+            return ErrorResponseUtil.errorResponse(new ConfigWritingDisabled());
+        }
+
+        final @NotNull Optional<DataCombiner> instance = combiningManager.getCombinerById(combiner.getId());
+        if (instance.isPresent()) {
+            return ErrorResponseUtil.errorResponse(new AlreadyExistsError(String.format(
+                    "DataCombiner already exists '%s'",
+                    combiner.getId())));
+        }
+
+
+        final DataCombiner dataCombiner = DataCombiner.fromModel(combiner);
+        try {
+            combiningManager.addDataCombiner(dataCombiner).get();
```

Comment:
here we call get to materialise the future, and updateCombiner does call get too, but it does it implicitly, ce n'est pas symétrique

Reply:
ce n'est pas symétrique -> what? :-D guess you mean it is not symmetric? Yeah, the methods are not fleshed out as the whole handling needs to be implemented. We do not have AsyncResponse, which means we have to sync the future at some point unfortunately. I wondered during the rough implementation whether async even makes sense then get() is called in the end.